### PR TITLE
First commit of HashMap + Watch

### PR DESCRIPTION
**Pull request by davidkurkov**
First working version of HashMap.
Also made a StopWatch class (known limitations right now, will fix).

### CODE REVIEW

**Comment by davidkurkov** [comment]
@goishin , there's definitely super ugly repetitive code going on in HashMap.java - this is the first implementation of it. I'll do cleanup once we review.

File: src/HashMap.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    int[] valueTable;
+    int size;
+    float loadFactor = 0.70f;
+    Prime primeNumberGenerator;
+
+    HashMap() {
+        clear();
+    }
+
+    @Override
```
Comment (goishin) [inline src/HashMap.java:17]: What happens if key is null from the caller's point of view? Do they have any way of verifying if the value was put in?

File: src/HashMap.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    public void put(String key, int value) {
+        if (key == null) {
+            return;
+        }
+        checkOverflow();
+        key = key.toLowerCase();
+        int index = hash(key, this.tableSize);
+        index = checkIndex(key, index, this.keyTable, this.tableSize);
```
Comment (goishin) [inline src/HashMap.java:25]: if(keyTable[index] != null && keyTable[index].compareTo(key) == 0)
{
    size -= 1;
}

File: src/hash.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+
+/**
+ * Created by david on 4/19/16.
+ */
+abstract interface hash {
+
```
Comment (goishin) [inline src/hash.java:6]: Need to have an overload for the [] operator to support usage like this:
hashmap['my key'] = 'some value'

File: src/hash.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    public void put(String key, int val); //** Add a new key-value pair to the map. If the key is already in the map then replace the old value with the new value. */
```
Comment (goishin) [inline src/hash.java:7]: Need to also support iterating through all values. Iterator should return a pair object (or tuple, up to you) with key being the first value in the pair and value being the second value.

File: src/HashMap.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            return false;
+        }
+        else if (keyTable[index].compareTo(key) == 0) {
+            return false;
+        }
+        return true;
+    }
+
+    private int checkIndex(String key, int index, String[] keyTable, int tableSize) {
+        if (indexTaken(key, index, keyTable)) {
```
Comment (goishin) [inline src/HashMap.java:125]: Don't make this recursive. Use a while loop. While you haven't found a spot, keep checking. If you encounter the same spot you started from, you need to change your strategy (such as recheck the load factor, etc.)

File: src/Watch.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    public void printResults() {
+        StringBuilder formattedResults = new StringBuilder();
+        for (String key: results.keys()) {
+            formattedResults.append(key);
+            formattedResults.append(" took ");
+            formattedResults.append(results.get(key));
+            formattedResults.append("s");
+            formattedResults.append("\n");
+        }
+        System.out.print(formattedResults);
```
Comment (goishin) [inline src/Watch.java:39]: System.out.print(formattedResults.toString());

**Pull request merged**